v3/src/components/graph/components/dotplotdots.tsx

```diff
@@ -1,102 +1,207 @@
+import {appState} from "../../app-state"
+import {useDataSetContext} from "../../../hooks/use-data-set-context"
 import {INumericAxisModel} from "../models/axis-model"
-import { useGraphLayoutContext } from "../models/graph-layout"
+import {useGraphLayoutContext} from "../models/graph-layout"
+import {ICase} from "../../../data-model/data-set"
+import {prf} from "../../../utilities/profiler"
+import {getScreenCoord, setPointCoordinates, setPointSelection} from "../utilities/graph_utils"
+import {reaction} from "mobx"
+import {onAction} from "mobx-state-tree"
+import {isSelectionAction, isSetCaseValuesAction} from "../../../data-model/data-set-actions"
 
 export const DotPlotDots = memo(observer(function DotPlotDots(props: {
   dots: plotProps,
   axisModel: INumericAxisModel,
   graphData: InternalizedData,
   dotsRef: React.RefObject<SVGSVGElement>
 }) {
-  const { graphData, dotsRef, axisModel } = props,
+  const {graphData, dotsRef, axisModel} = props,
     dataset = useDataSetContext(),
     layout = useGraphLayoutContext(),
-    xScale = layout.axisScale(axisModel.place),
-    [dragID, setDragID] = useState<string>(),
-    [, setRefreshCounter] = useState(0),
+    place = axisModel.place,
+    countPlace = place === 'left' ? 'bottom' : 'left',
+    xScale = layout.axisScale(place),
+    [rangeMin, rangeMax] = xScale?.range() || [],
+    plotWidth = Math.abs(rangeMax ?? 0 - rangeMin ?? 0),
+    yScale = layout.axisScale(countPlace),
+    [dragID, setDragID] = useState(''),
     currPos = useRef({x: 0}),
+    didDrag = useRef(false),
     target = useRef<any>(),
-    [firstTime, setFirstTime] = useState<boolean | null>(true),
+    firstTime = useRef(true),
+    xAttrID = graphData.xAttributeID,
     selectedDataObjects = useRef<Record<string, { x: number }>>({})
 
   const onDragStart = useCallback((event: MouseEvent) => {
       appState.beginPerformance()
-      if (firstTime) {
-        setFirstTime(false) // We never want to animate points on drag
-      }
-      const xAttrID = graphData.xAttributeID
+      dataset?.beginCaching()
+      firstTime.current = false // We don't want to animate points until end of drag
+      didDrag.current = false
       target.current = select(event.target as SVGSVGElement)
       const tItsID: string = target.current.property('id')
       if (target.current.node()?.nodeName === 'circle') {
         target.current.transition()
           .attr('r', dragRadius)
         setDragID(() => tItsID)
         currPos.current = {x: event.clientX}
-        dataset?.selectCases([tItsID])
+
+        const [, caseId] = tItsID.split("_")
+        dataset?.selectCases([caseId])
         // Record the current values so we can change them during the drag and restore them when done
         dataset?.selection.forEach(anID => {
           selectedDataObjects.current[anID] = {
-            x: dataset?.getNumeric(anID, xAttrID) ?? NaN
+            x: dataset?.getNumeric(anID, xAttrID) ?? 0
```

Comment:
It can't get into the selection without having a value. But, in spirit of knowing that whatever can go wrong will, I'm going to prevent it from getting in in the unlikely event that it doesn't have a value.